src/utils/arr.h

```diff
@@ -73,13 +75,44 @@ static array_t array_new_sz(uint32_t elem_sz, uint32_t cap, uint32_t len) {
  *  */
 #define array_new(T, cap) (array_new_sz(sizeof(T), cap, 0))
 
+#define array_new_on_stack(T, capacity, varName)    \
+  struct {                                          \
+    bool on_stack;                                  \
+    uint32_t len;                                   \
+    uint32_t cap;                                   \
+    uint32_t elem_sz;                               \
+    char buf[(capacity) * sizeof(T)];               \
+  } varName ## _struct;                             \
+  varName ## _struct.on_stack = true;               \
+  varName ## _struct.len = 0;               \
+  varName ## _struct.cap = (capacity);              \
+  varName ## _struct.elem_sz = sizeof(T);           \
+  T* varName = varName ## _struct.buf;
+
 /* Initialize an array for a given type T with a given length. The capacity allocated is identical
  * to the length
  *  */
 #define array_newlen(T, len) (array_new_sz(sizeof(T), len, len))
 
+static inline array_t persist(array_t arr){
+  array_hdr_t *hdr = array_hdr(arr);
+  if(!hdr->on_stack) return arr;
+  array_t tmp = array_new_sz(hdr->elem_sz, hdr->cap, hdr->len);
+  memcpy(tmp, arr, hdr->cap);
```

Comment:
shouldn't we just use adr->len ? also on like above? There is no need to allocate the entire capacity? what do you think?